File: 03_02_MinStack.java
// Solution 1

class MinStack {
    Deque<Integer> stack;
    Deque<Integer> minValue;

    public MinStack() {
        stack = new ArrayDeque<>();
        minValue = new ArrayDeque<>();
    }
    
    public void push(int val) {
        stack.push(val);
        
        if (minValue.isEmpty() || val <= minValue.peek()) {
            minValue.push(val);
        }
    }
    
    public void pop() {
        int val = stack.pop();
        
        if (val == minValue.peek()) {
             minValue.pop();
        }
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minValue.peek();
    }
}


// Solution 2

class MinStack {
    Deque<Integer> stack;

    public MinStack() {
        stack = new ArrayDeque<>();
        
    }
    
    public void push(int val) {
        stack.push(val);
    }
    
    public void pop() {
        stack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        int minValue = stack.peek();
        
        for (int val : stack) {
            minValue = Math.min(minValue, val);
        }
        return minValue;
    }
}

File: 03_02_MinStack.py
class MinStack:
    """
    MinStack implementation
    This solution uses two stacks, one holding the items on stack and the 
    other storing the current minimum 
    ie. stack = [5, 3, 4, 2, 3, 4, 1]
        mins =  [5, 3, 3, 2, 2, 2, 1]
        Push -> we append to the stack and also check with the top of the min stack
        for the current min which is stored in the mins stack
        Pop -> We pop from both the stack and mins
        getMin -> We retrieve the top element on the min stack
    """

    def __init__(self):
        self.stack = []
        self.mins = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if len(self.mins) == 0:
            self.mins.append(val)
            return
        self.mins.append(min(self.mins[-1], val))
        
        
    def pop(self) -> None:
        self.mins.pop()
        return self.stack.pop()
        

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.mins[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
